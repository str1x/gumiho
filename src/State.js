import {
    fromJS,
    isImmutable,
    getIn,
} from 'immutable';

const validPropTypes = ['number', 'string', 'boolean'];

export default class State {

    constructor(rawState = {}) {
        this._state = State.normalize(rawState);
    }

    get(...args) {
        const value = getIn(this._state, args);
        return isImmutable(value) ? value.toJS() : value;
    }

    set() {

    }

    static normalize(rawState) {
        return fromJS(rawState, (key, value) => {
            console.log(value);
            if (isImmutable(value)) {
                return value.filter( i => {
                    return validPropTypes.includes(typeof i) || isImmutable(i);
                });
            }
            if (validPropTypes.includes(typeof value)) {
                return value;
            }
            return undefined;
        });
    }
}
