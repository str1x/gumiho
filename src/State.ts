type PureType = string|number|boolean|undefined|null;
interface FlatState {
    [key: string]: PureType;
};
interface Config {
    pathSeparator: string;
    typeSeparator: string;
}
const pureTypes = ['string', 'number', 'boolean'];

export default class State {
    private _state: FlatState | PureType = {};
    private _config: Config = {
        pathSeparator: '/',
        typeSeparator: '#',
    };

    constructor(rawState: unknown) {
        const {state} = State.makeFlatState(rawState);
        this._state = state;
    }

    static isPure(value: unknown): boolean {
        if (
            pureTypes.includes(typeof value)
            || value === undefined
            || value === null
        ) {
            value as PureType;
            return true;
        }
        return false;
    }

    static makeFlatState(
        rawState: unknown | {[key: string]: unknown}, 
        rootState: FlatState = {}, 
        pathSeparator = '/', 
        typeSeparator = '#', 
        pathTo: Array<string> = []
    ): {
        state: FlatState | PureType; 
        statePath: Array<string> | null;
    } {
        let type: 'array' | 'object' | null = null;
        
        if (State.isPure(rawState)) {
            const state = rawState as PureType;
            
            return {
                state,
                statePath: pathTo,
            };
        }
        else if (Array.isArray(rawState)) {
            type = 'array';
        }
        else if (rawState.toString() === '[object Object]') {
            type = 'object';
        }

        if (type === null) {
            return {
                state: undefined,
                statePath: pathTo,
            };
        }
        else {
            let path = [...pathTo];
            path[path.length - 1] = `${type}${typeSeparator}${path[path.length - 1]}`;
            if (!path.length) {
                path = [`${type}${typeSeparator}`];
            }
            const keys = Object.keys(rawState);
            keys.forEach((key) => {
                const {state, statePath} = State.makeFlatState(
                    (rawState as {[key: string]: unknown})[key],
                    rootState,
                    pathSeparator,
                    typeSeparator,
                    [...path, key],
                );
                if (statePath.length) {
                    rootState[statePath.join(pathSeparator)] = state as PureType;
                }
            });
            return {
                state: rootState,
                statePath: [],
            };
        }
    }
}
