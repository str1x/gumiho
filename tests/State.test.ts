import State from '@src/State';

function stringifyState(state: unknown): string {
    return JSON.stringify(state, undefined, '    ');
}

describe('State', () => {
    const pureValues = ['state string', 666, 666.666, true, false, undefined, NaN, null];
    const validEntities = [[], {}];
    const notValidEntities = [Symbol('sym'), /w/, () => {}];
    const entities = [...validEntities, ...notValidEntities];

    describe('isPure', () => {
        const { isPure } = State;
        it('should return false for pure values', () => {
            for (const value of pureValues) {
                expect(isPure(value)).toBe(true);
            }
        });
        it('should return false for entities', () => {
            for (const value of entities) {
                expect(isPure(value)).toBe(false);
            }
        });
    });

    describe('makeFlatState', () => {
        const { makeFlatState } = State;
        it ('with pure', () => {
            for (const value of pureValues) {
                const flatState = makeFlatState(value);
                expect(flatState).toEqual({
                    state: value,
                    statePath: []
                });
            }
        });
        it('should create state for with valid entities', () => {
            for (const value of validEntities) {
                const flatState = makeFlatState(value);
                expect(flatState).toEqual({
                    state: {},
                    statePath: [],
                });
            }
        });
        it('should create state with undefined for not valid entities', () => {
            for (const value of notValidEntities) {
                const flatState = makeFlatState(value);
                expect(flatState).toEqual({
                    state: undefined,
                    statePath: [],
                });
            }
        });
        it('should create state for nested object in array', () => {
            const rawState = [1, '2', false, /w/, [3],
                {
                    propA: 1,
                    propB: 'some prop',
                    propC: true,
                    propD: () => {},
                    propE: ['1'],
                    propF: {
                        someProp: 1,
                    },
                }
            ];
            const flatState = makeFlatState(rawState);
            expect(Object.keys(flatState.state)).toHaveLength(11);
            expect(stringifyState(flatState)).toMatchSnapshot();
        });
    });

    describe('get', () => {
        it('shoult return correct value for pure field', () => {
            for (const value of pureValues) {
                const state = new State(value);
                expect(state.get() === value)
            }
        });
        it('should return empty entites', () => {
            for (const value of validEntities) {
                const state = new State(value);
                expect(state.get() === {});
            }
        });
        it('should return undefined for not valid entites', () => {
            for (const value of notValidEntities) {
                const state = new State(value);
                expect(state.get() === undefined);
            }
        });
    });
});
