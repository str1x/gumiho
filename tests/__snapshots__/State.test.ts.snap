// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State makeFlatState should create state for nested object in array 1`] = `
"{
    \\"state\\": {
        \\"array#/0\\": 1,
        \\"array#/1\\": \\"2\\",
        \\"array#/2\\": false,
        \\"array#/array#4/0\\": 3,
        \\"array#/object#5/propA\\": 1,
        \\"array#/object#5/propB\\": \\"some prop\\",
        \\"array#/object#5/propC\\": true,
        \\"array#/object#5/array#propE/0\\": \\"1\\",
        \\"array#/object#5/object#propF/someProp\\": 1
    },
    \\"statePath\\": []
}"
`;
